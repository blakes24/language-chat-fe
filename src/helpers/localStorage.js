export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
