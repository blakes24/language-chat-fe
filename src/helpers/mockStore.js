import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
