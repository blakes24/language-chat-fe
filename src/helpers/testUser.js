import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
