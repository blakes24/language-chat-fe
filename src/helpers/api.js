import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
