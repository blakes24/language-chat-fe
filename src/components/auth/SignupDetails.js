import { useFormik, FormikProvider } from "formik";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import FormikSelect from "../FormikSelect";
import { useHistory } from "react-router-dom";
import ChatApi from "../../helpers/api";
import { useDispatch } from "react-redux";
import { setToken } from "../../store/usersSlice";
import { setLocalStorage } from "../../helpers/localStorage";

const validate = (values) => {
  const errors = {};
  if (!values.name) {
    errors.name = "Required";
  }
  if (!values.bio) {
    errors.bio = "Required";
  }
  return errors;
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    flexWrap: "wrap",
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: "25ch",
  },
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
}));

function SignupDetails({
  email,
  name,
  password,
  imageUrl,
  socialProvider,
  socialId,
}) {
  const history = useHistory();
  const dispatch = useDispatch();

  const formik = useFormik({
    initialValues: {
      email: email,
      name: name,
      password: password,
      imageUrl: imageUrl,
      socialProvider: socialProvider,
      socialId: socialId,
      bio: "",
      speaksLang: "",
      learnsLang: "",
      learnsLevel: "",
    },
    validate,
    onSubmit: async (values) => {
      try {
        const res = await ChatApi.register(values);
        dispatch(setToken(res));
        setLocalStorage("token", res);
        history.push("/");
      } catch (err) {
        formik.errors.name = err;
      }
    },
  });

  const languages = [
    { label: "Arabic", value: "ar" },
    { label: "Chinese", value: "zh" },
    { label: "English", value: "en" },
    { label: "French", value: "fr" },
    { label: "German", value: "de" },
    { label: "Hindi", value: "hi" },
    { label: "Indonesian", value: "id" },
    { label: "Italian", value: "it" },
    { label: "Japanese", value: "ja" },
    { label: "Korean", value: "ko" },
    { label: "Portuguese", value: "pt" },
    { label: "Russian", value: "ru" },
    { label: "Spanish", value: "es" },
  ];
  const levels = [
    { value: 1, label: "Beginner" },
    { value: 2, label: "Intermediate" },
    { value: 3, label: "Advanced" },
  ];
  const classes = useStyles();

  return (
    <FormikProvider value={formik}>
      <form
        className={classes.root}
        autoComplete="off"
        onSubmit={formik.handleSubmit}
      >
        <TextField
          id="name"
          name="name"
          label="Name"
          variant="outlined"
          style={{ margin: 8 }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.name}
          FormHelperTextProps={{ error: true }}
          helperText={
            formik.errors.name && formik.touched.name && formik.errors.name
          }
        />
        <TextField
          id="bio"
          name="bio"
          label="About Me"
          multiline
          rows={4}
          variant="outlined"
          style={{ margin: 8 }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.bio}
          FormHelperTextProps={{ error: true }}
          helperText={
            formik.errors.bio && formik.touched.bio && formik.errors.bio
          }
        />
        <FormikSelect
          name="speaksLang"
          items={languages}
          label="I speak"
          instruction="Select a Language"
          className={classes.formControl}
          value={formik.values.speaksLang}
          helperText="hi"
          required
        />
        <FormikSelect
          name="learnsLang"
          items={languages}
          label="I am learning"
          instruction="Select a Language"
          className={classes.formControl}
          value={formik.values.learnsLang}
          required
        />
        <FormikSelect
          name="learnsLevel"
          items={levels}
          label="My level is"
          instruction="Select a Level"
          className={classes.formControl}
          value={formik.values.learnsLevel}
          required
        />
        <Button
          variant="contained"
          color="secondary"
          type="submit"
          fullWidth
          size="large"
          style={{ margin: 8 }}
        >
          Create Account
        </Button>
      </form>
    </FormikProvider>
  );
}

export default SignupDetails;
