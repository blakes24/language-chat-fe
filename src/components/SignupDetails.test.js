import { render } from "@testing-library/react";
import SignupDetails from "./SignupDetails";
import { UserProvider } from "../helpers/testUser";

test("renders without crashing", () => {
  render(
    <UserProvider>
      <SignupDetails />
    </UserProvider>
  );
}); 

it("matches snapshot", function () {
  const { asFragment } = render(
    <UserProvider>
      <SignupDetails />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});