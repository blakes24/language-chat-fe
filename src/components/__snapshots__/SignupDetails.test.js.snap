// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="makeStyles-root-8"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style="margin: 8px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="name"
        id="name-label"
      >
        Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="name"
          name="name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-13"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style="margin: 8px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="bio"
        id="bio-label"
      >
        About Me
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="bio"
          name="bio"
          rows="4"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-13"
          >
            <span>
              About Me
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        style="margin: 8px;"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          I speak
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-speaksLang"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            id="mui-component-select-speaksLang"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="speaksLang"
            required=""
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-13"
            >
              <span>
                I speak *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        style="margin: 8px;"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          I am learning
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-learnsLang"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            id="mui-component-select-learnsLang"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="learnsLang"
            required=""
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-13"
            >
              <span>
                I am learning *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
        style="margin: 8px;"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          My level is
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-learnsLevel"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            id="mui-component-select-learnsLevel"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="learnsLevel"
            required=""
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-13"
            >
              <span>
                My level is *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
      style="margin: 8px;"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Create Account
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
